common/blobstore/s3store/client.go

```diff
@@ -0,0 +1,311 @@
+		}
+	}()
+
+	body, err := ioutil.ReadAll(result.Body)
+
+	return body, err
+}
+
+func s3gettags(ctx context.Context, s3api s3iface.S3API, bucket string, key string) (map[string]string, error) {
+
+	result, err := s3api.GetObjectTaggingWithContext(ctx, &s3.GetObjectTaggingInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	tags := make(map[string]string)
+	for _, e := range result.TagSet {
+
+		tags[*e.Key] = *e.Value
+
+	}
+	return tags, err
+}
+func (c *client) Upload(ctx context.Context, bucket string, key blob.Key, blob *blob.Blob) error {
+	params := url.Values{}
+	for k, v := range blob.Tags {
+		params.Add(k, v)
+	}
+
+	return s3upload(ctx, c.s3cli, bucket, key.String(), blob.Body, params)
+}
+
+func (c *client) Download(ctx context.Context, bucket string, key blob.Key) (*blob.Blob, error) {
+
+	b, err := s3download(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		if err == ErrReadFile {
+			return nil, blobstore.ErrBlobNotExists
+		}
+		return nil, err
+	}
+	tags, err := s3gettags(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		return nil, err
+	}
+	return blob.NewBlob(b, tags), nil
+}
+
+func (c *client) GetTags(ctx context.Context, bucket string, key blob.Key) (map[string]string, error) {
+
+	tags, err := s3gettags(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		return nil, err
+	}
+	return tags, nil
+}
+
+func (c *client) Exists(ctx context.Context, bucket string, key blob.Key) (bool, error) {
+
+	_, err := c.s3cli.HeadObjectWithContext(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		aerr, ok := err.(awserr.Error)
+		if ok && aerr.Code() == s3.ErrCodeNoSuchKey {
+			return false, nil
+		}
+		if ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return false, blobstore.ErrBucketNotExists
+		}
+
+		return false, err
+	}
+	return true, nil
+}
+
+func (c *client) Delete(ctx context.Context, bucket string, key blob.Key) (bool, error) {
+
+	_, err := c.s3cli.DeleteObjectWithContext(ctx, &s3.DeleteObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return false, blobstore.ErrBucketNotExists
+			}
+
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				return false, nil
+			}
+
+		}
+		return false, err
+
+	}
+	return true, nil
+}
+
+func (c *client) ListByPrefix(ctx context.Context, bucket string, prefix string) ([]blob.Key, error) {
+
+	results, err := c.s3cli.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, ErrListFiles
+	}
+	var keys = make([]blob.Key, len(results.Contents))
```

Comment:
Since you actually know the number of items that will be in this slice you can declare it with both length and capacity. This is more efficient (e.g. make([]int, 5, 5)